### PR TITLE
Fix more bugs found in fuzzing

### PR DESCRIPTION
**Pull request by joshuawarner32**
- **Fix nasty perf bug in parsing types**
- **Add some currently-passing tests**
- **Fix round-trip parse->fmt->parse for dbg stmts with more than one arg**
- **Improve normalization to handle cases found in fuzzing**
- **Improve parsing of negative + not'd exprs**
- **Fix a couple minor parsing bugs**
- **Fix a bunch of bugs found in fuzzing**
- **Several fixes found in fuzzing**
- **Record updates**
- **Fix fuzz Cargo.toml regression**
- **Adjust parens around patterns in annotations**
- **Fix indent calls in optional fields**
- **Unify if/when condition formatting**
- **Fix newline after dbg formatting**
- **Fix purity removal**
- **Fix ability with no newline after it**
- **Fix indent call in when fmt**
- **Fix multiline pattern in body, which may only apply to multiline strings**
- **Fix lifting of backpassing**
- **Make sure patterns continue to parse as such**
- **Add lifting for ext values in types**
- **Only allow indented else if there's a newline after the else**
- **Don't do empty record destructure elision when the body is itself Defs**
- **Fix parsing crash with dollar escapes**
- **Make parsing of conditionals consistent between stmts/exprs**
- **Disallow conditionals after neg/not**
- **Fix expr_lift_spaces of Closure**
- **Make non-equivalent annotation+body pairs not be merged**
- **Fix formatting of newlines in parens in a function type**
- **Indented seq parsing for imports**
- **Flesh out expr_lift_spaces, in particular handling DbgStmt properly**
- **Make sure ext's can't slurp surrounding apply args**
- **Fixup negative allowance in calls**
- **Ensure formatted multiline patterns can be parsed**
- **Fixup string dollar handling**
- **Fix warnings**

### CODE REVIEW

**Comment by joshuawarner32** [comment]
#7307 will make it much easier to fix the conflicts here, so we should get that one in first, then I'll rebase this.

**Review by smores56 — COMMENTED** [review]
Only just started reading, I'll get back to this later this week

File: crates/compiler/test_syntax/fuzz/Cargo.toml [hunk 1]
```diff
@@ -10,9 +10,9 @@ edition = "2021"
 cargo-fuzz = true
 
 [dependencies]
-test_syntax.workspace = true
-roc_parse.workspace = true
-bumpalo = { workspace = true, features = ["collections"] }
+test_syntax = { path = "../../test_syntax" }
+roc_parse = { path = "../../parse" }
+bumpalo = { version = "3.12.0", features = ["collections"] }
```
[the remark this answers is not in the document]
Comment (joshuawarner32) [reply crates/compiler/test_syntax/fuzz/Cargo.toml:15]: I can't - the fuzz cargo.toml is part of a separate workspace (intentionally so!). This is undoing a change that broke fuzzing.

File: crates/compiler/fmt/src/spaces.rs [hunk 2]
```diff
@@ -20,6 +20,38 @@ pub fn fmt_default_newline(buf: &mut Buf, spaces: &[CommentOrNewline], indent: u
...
+    buf: &mut Buf<'_>,
+    mut spaces: &[CommentOrNewline<'_>],
+    indent: u16,
+    mode: SpacesNewlineMode,
+) {
+    if matches!(
+        mode,
+        SpacesNewlineMode::SkipNewlinesAtStart | SpacesNewlineMode::SkipNewlinesAtBoth
+    ) {
+        while let Some(CommentOrNewline::Newline) = spaces.first() {
```
[the remark this answers is not in the document]
Comment (joshuawarner32) [reply crates/compiler/fmt/src/spaces.rs:40]: FWIW, it's not clear to me a-priori if there's any meaningful difference between those from a perf perspective. I agree that's a bit more readable perhaps.

File: crates/compiler/test_syntax/fuzz/Cargo.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (smores56) [reply crates/compiler/test_syntax/fuzz/Cargo.toml:15]: Can you leave a comment that says we shouldn't try to merge this in the future?

[the remark this answers is not in the document]
Comment (smores56) [reply crates/compiler/test_syntax/fuzz/Cargo.toml:15]: I guess there's one on line 18, but apparently it's not visible enough

[the remark this answers is not in the document]
Comment (joshuawarner32) [reply crates/compiler/test_syntax/fuzz/Cargo.toml:15]: Good idea.

Also, I'm hoping to get a short round of fuzzing (say, 60 seconds) checked in as a CI job shortly-ish, which should catch easy to hit bugs as well as any problems in the fuzzing target itself.

[the remark this answers is not in the document]
Comment (smores56) [reply crates/compiler/test_syntax/fuzz/Cargo.toml:18]: That'd be great!

File: crates/compiler/fmt/src/expr.rs [hunk 3]
```diff
@@ -29,311 +31,308 @@ impl<'a> Formattable for Expr<'a> {
...
+            buf.indent(indent);
+            buf.push_str("crash");
+        }
+        Expr::Try => {
+            buf.indent(indent);
+            buf.push_str("try");
+        }
+        Expr::Apply(loc_expr, loc_args, _) => {
+            let apply_needs_parens = parens == Parens::InApply || parens == Parens::InApplyLastArg;
+            // || (parens == Parens::InOperator && apply_has_closure(self));
```
Comment (lukewilliamboswell) [inline crates/compiler/fmt/src/expr.rs:93]: Should this be removed? maybe a follow up

File: crates/compiler/fmt/src/expr.rs [hunk 4]
```diff
@@ -1832,41 +1922,71 @@ fn fmt_record_like<'a, Field, Format, ToSpaceBefore>(
...
-                        buf.newline();
-                    }
+                // if index == 0
+                //     && field_lifted
+                //         .before
+                //         .first()
+                //         .map_or(false, |s| s.is_comment())
+                // {
+                //     buf.ensure_ends_with_newline();
+                // }
```
Comment (lukewilliamboswell) [inline crates/compiler/fmt/src/expr.rs:1947]: should this be removed?

File: crates/compiler/fmt/src/expr.rs [hunk 5]
```diff
@@ -1832,41 +1922,71 @@ fn fmt_record_like<'a, Field, Format, ToSpaceBefore>(
...
+                    } else {
+                        SpacesNewlineMode::SkipNewlinesAtStart
+                    }
+                } else {
+                    SpacesNewlineMode::Normal
+                };
+
+                fmt_spaces_with_newline_mode(buf, before, field_indent, newline_mode);
+                // buf.ensure_ends_with_newline();
+                // fmt_comments_only(buf, before.iter(), NewlineAt::Bottom, field_indent);
```
Comment (lukewilliamboswell) [inline crates/compiler/fmt/src/expr.rs:1965]: should this be removed?

**Review by lukewilliamboswell — APPROVED** [review]

**Review by lukewilliamboswell — APPROVED** [review]

**Pull request merged**